File: AirHack/PPSClient.m
//
//  PPSClient.m
//  AirHack
//
//  Created by banjun on 2014/07/13.
//  Copyright (c) 2014年 banjun. All rights reserved.
//

#import "PPSClient.h"
#import "PPSLocalSong.h"
#import <AFNetworking.h>
#import "Functional.h"


static NSString * const kPost = @"POST";
static NSString * const kSongsAdd = @"/songs/add";
static NSString * const kSongsSkip = @"/songs/skip";
static NSString * const kSongsIndexHTML = @"/songs/index.html";
static NSString * const kParamFile = @"file";
static NSString * const kParamTitle = @"title";
static NSString * const kParamArtist = @"artist";
static NSString * const kParamArtwork = @"artwork";
static NSString * const kContentTypeOctetStream = @"application/octet-stream";
static NSString * const kContentTypeJpeg = @"image/jpeg";


@interface PPSClient ()

@property (nonatomic) NSURL *baseURL;

@property (nonatomic, copy) void (^currentProgressHandler)(float progress);

@end


@implementation PPSClient

- (instancetype)initWithBaseURL:(NSURL *)url;
{
    if (self = [super init]) {
        self.baseURL = url;
    }
    return self;
}

- (void)pushSongs:(NSArray *)songs progress:(void (^)(float progress))progressHandler didPushSong:(void (^)(PPSLocalSong *song))didPushSong completion:(void (^)())completion failure:(void (^)(NSError *error))failure; // Array<PPSSong>
{
    self.currentProgressHandler = progressHandler;
    
    AFURLSessionManager *manager = [[AFURLSessionManager alloc] initWithSessionConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    manager.responseSerializer = [[AFHTTPResponseSerializer alloc] init];
    
    for (PPSLocalSong *song in songs) {
        NSMutableURLRequest *request = [[AFHTTPRequestSerializer serializer] multipartFormRequestWithMethod:kPost URLString:[self.baseURL.absoluteString stringByAppendingString:kSongsAdd] parameters:nil constructingBodyWithBlock:^(id<AFMultipartFormData> formData) {
            // song file content
            [formData appendPartWithFileData:[NSData dataWithContentsOfFile:song.filePath] name:kParamFile fileName:song.filePath.lastPathComponent mimeType:kContentTypeOctetStream];
            
            // metadata
            void (^addText)(NSString *, NSString *) = ^(NSString *key, NSString *value) {
                [formData appendPartWithFormData:[[value stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding] dataUsingEncoding:NSUTF8StringEncoding]
                                            name:[key stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding]];
            };
            addText(kParamTitle, song.title);
            addText(kParamArtist, song.artist);
            [formData appendPartWithFileData:song.artworkJPEGData name:kParamArtwork fileName:kParamArtwork mimeType:kContentTypeJpeg];
        } error:nil];
        
        NSProgress *progress = nil;
        NSURLSessionUploadTask *uploadTask = [manager uploadTaskWithStreamedRequest:request progress:&progress completionHandler:^(NSURLResponse *response, id responseObject, NSError *error) {
            
            NSHTTPURLResponse *res = (NSHTTPURLResponse *)response;
            NSInteger status = res.statusCode;
            if (error || status != 200) {
                if (failure) {
                    failure(error);
                }
            } else {
                song.uploadCompleted = YES;
                if (didPushSong) {
                    didPushSong(song);
                }
                BOOL hasAllDone = [songs all:^BOOL(PPSLocalSong *s){ return s.uploadCompleted; }];
                if (hasAllDone && completion) {
                    completion();
                }
            }
        }];
        song.uploadProgress = progress;
        [uploadTask resume];
    }
}

- (void)skip
{
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
    manager.responseSerializer = [[AFHTTPResponseSerializer alloc] init];
    [manager POST:[self.baseURL.absoluteString stringByAppendingString:kSongsSkip] parameters:nil success:^(AFHTTPRequestOperation *operation, id responseObject) {
        NSLog(@"skip succeeded");
    } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
        NSLog(@"skip failed: %@", error);
    }];
}

- (NSURL *)songsIndexHTMLURL
{
    return [NSURL URLWithString:kSongsIndexHTML relativeToURL:self.baseURL];
}

@end

File: AirHack/PPSClient.swift
//
//  PPSClient.swift
//  AirHack
//
//  Created by BAN Jun on 2014/09/21.
//  Copyright (c) 2014年 banjun. All rights reserved.
//

import Foundation
import MediaPlayer

let PPSClientNowPlayingDidChangeNotification: NSString = "PPSClientNowPlayingDidChangeNotification"

private let kPost = "POST"
private let kSongsAdd = "/songs/add"
private let kSongsSkip = "/songs/skip"
private let kSongsIndexHTML = "/songs/index.html"
private let kSongsIndexJSON = "/songs/index.json"
private let kParamFile = "file"
private let kParamTitle = "title"
private let kParamArtist = "artist"
private let kParamArtwork = "artwork"
private let kContentTypeOctetStream = "application/octet-stream"
private let kContentTypeJpeg = "image/jpeg"
private let kNSProgressCompletedKey = "fractionCompleted"


class PPSClient : NSObject {
    let baseURL: NSURL
    var nowPlaying: PPSNowPlaying? {
        didSet {
            NSNotificationCenter.defaultCenter().postNotificationName(PPSClient.nowPlayingDidChangeNotificationName, object: self)
        }
    }
    let sessionManager: AFURLSessionManager
    var songIDToProgressMap: [(LocalSong, NSProgress)] = []
    
    class var nowPlayingDidChangeNotificationName : String { return "PPSClientNowPlayingDidChangeNotification" }
    
    var songsIndexHTMLURL: NSURL {
        return baseURL.URLByAppendingPathComponent(kSongsIndexHTML)
    }
    
    init(baseURL: NSURL) {
        let sessionConfiguration = NSURLSessionConfiguration.defaultSessionConfiguration()
        sessionConfiguration.HTTPMaximumConnectionsPerHost = 1
        sessionManager = AFURLSessionManager(sessionConfiguration: sessionConfiguration)
        sessionManager.responseSerializer = AFHTTPResponseSerializer()
        self.baseURL = baseURL
    }
    
    func songsAdd(song: LocalSong, completion: (() -> Void)?) {
        if song.exportFilePath == nil { completion?(); return; }
        let filePath = song.exportFilePath!
        
        song.status = .Posting(0)
        
        let serializer = AFHTTPRequestSerializer()
        let request = serializer.multipartFormRequestWithMethod(kPost, URLString: baseURL.URLByAppendingPathComponent(kSongsAdd).absoluteString, parameters: nil, constructingBodyWithBlock: { (formData: AFMultipartFormData!) -> Void in
            // song file content
            formData.appendPartWithFileData(NSData(contentsOfFile: filePath), name: kParamFile, fileName: filePath.lastPathComponent, mimeType: kContentTypeOctetStream)
            
            // metadata
            func addText(key: String, value: String) -> Void {
                formData.appendPartWithFormData(value.stringByAddingPercentEscapesUsingEncoding(NSUTF8StringEncoding)?.dataUsingEncoding(NSUTF8StringEncoding, allowLossyConversion: true),
                    name: key.stringByAddingPercentEscapesUsingEncoding(NSUTF8StringEncoding))
            }
            addText(kParamTitle, song.mediaItem.title ?? "no title")
            addText(kParamArtist, song.mediaItem.artist ?? "unknown artist")
            formData.appendPartWithFileData(song.artworkJPEGData, name: kParamArtwork, fileName: kParamArtwork, mimeType: kContentTypeJpeg)
            }, error: nil)
        
        var progress: NSProgress? = nil
        let uploadTask = sessionManager.uploadTaskWithStreamedRequest(request, progress: &progress) {
            (response: NSURLResponse!, responseObject: AnyObject!, error: NSError!) -> Void in
            let status = (response as? NSHTTPURLResponse)?.statusCode ?? 0
            if error != nil || status != 200 {
                song.status = .Failed("Send: \(error)")
            } else {
                song.status = .Posted
            }
            
            dispatch_async(dispatch_get_main_queue()) {
                if let p = progress {
                    p.removeObserver(self, forKeyPath: kNSProgressCompletedKey)
                    for i in 0...self.songIDToProgressMap.count {
                        if self.songIDToProgressMap[i].0 === song {
                            self.songIDToProgressMap.removeAtIndex(i)
                            break
                        }
                    }
                }
                completion?()
            }
        }
        if let p = progress {
            p.addObserver(self, forKeyPath: kNSProgressCompletedKey, options: NSKeyValueObservingOptions.allZeros, context: nil)
            songIDToProgressMap += [(song, p)]
        }
        
        uploadTask.resume()
    }
    
    override func observeValueForKeyPath(keyPath: String!, ofObject object: AnyObject!, change: [NSObject : AnyObject]!, context: UnsafeMutablePointer<Void>) {
        if object is NSProgress {
            for (song, progress) in songIDToProgressMap {
                if progress === object {
//                    println("progress: \(progress.fractionCompleted) for \(song)")
                    song.status = .Posting(Float(progress.fractionCompleted))
                    return
                }
            }
            return
        }
        super.observeValueForKeyPath(keyPath, ofObject: object, change: change, context: context)
    }
    
    func loadNowPlaying() {        
        let manager = AFHTTPRequestOperationManager()
        manager.responseSerializer = AFJSONResponseSerializer()
        manager.GET(baseURL.URLByAppendingPathComponent(kSongsIndexJSON).absoluteString, parameters: nil, success: { (operation: AFHTTPRequestOperation!, responseObject: AnyObject!) -> Void in
            if let np = PPSNowPlaying(JSON: responseObject as? Dictionary) as PPSNowPlaying? {
                self.nowPlaying = np
            } else {
                println("cannot get nowPlaying from json: \(responseObject)")
            }
        }, failure: { (_:AFHTTPRequestOperation!, error:NSError!) -> Void in
            println("\(kSongsIndexJSON) failed: \(error)")
        })
    }
    
    func skip() {
        let manager = AFHTTPRequestOperationManager()
        manager.responseSerializer = AFJSONResponseSerializer()
        manager.POST(baseURL.URLByAppendingPathComponent(kSongsSkip).absoluteString, parameters: nil,
            success: {(_: AFHTTPRequestOperation!, _: AnyObject!) in },
            failure: {(_:AFHTTPRequestOperation!, _:NSError!) in })
    }
}